### ISSUE TITLE
On failed connections remote_conn_pre attribute should be cleaned-up / deleted

### ISSUE THREAD

**Issue by ktbyers** [ktbyers/netmiko#826]
You can run into issues where VTY connections are consumed on a device because the connection has to wait for the VTY timeout to expire on failed authentications.

**Comment by jcrotts** [comment]
Could this be accomplished by adding .close() calls before the error message gets raised in establish_connection?
`# initiate SSH connection
            try:
                self.remote_conn_pre.connect(**ssh_connect_params)
            except socket.error:
                msg = "Connection to device timed-out: {device_type} {ip}:{port}".format(
                    device_type=self.device_type, ip=self.host, port=self.port)
                self.remote_conn_pre.close() #
                raise NetMikoTimeoutException(msg)
            except paramiko.ssh_exception.AuthenticationException as auth_err:
                msg = "Authentication failure: unable to connect {device_type} {ip}:{port}".format(
                    device_type=self.device_type, ip=self.host, port=self.port)
                msg += self.RETURN + str(auth_err)
                self.remote_conn_pre.close() #
                raise NetMikoAuthenticationException(msg)
`

**Comment by rismanchicm** [comment]
i just use self.disconnect() right before the auth exception is raised. I dont think its needed for timeout expection other than to set remote_conn to None because the connection is already gone? 

```python
try:
	self.remote_conn_pre.connect(**ssh_connect_params)
except socket.error:
	msg = "Connection to device timed-out: {device_type} {ip}:{port}".format(
		device_type=self.device_type, ip=self.host, port=self.port)
	raise NetMikoTimeoutException(msg)
except paramiko.ssh_exception.AuthenticationException as auth_err:
	msg = "Authentication failure: unable to connect {device_type} {ip}:{port}".format(
		device_type=self.device_type, ip=self.host, port=self.port)
	msg += self.RETURN + str(auth_err)
	self.disconnect()
	raise NetMikoAuthenticationException(msg)

```
Also dont forget about telnet_logon and the overrided one in cisco base
```python
def telnet_login(self, pri_prompt_terminator=r'#\s*$', alt_prompt_terminator=r'>\s*$',
				 username_pattern=r"(?:[Uu]ser:|sername|ogin)", pwd_pattern=r"assword",
				 delay_factor=1, max_loops=20):
	"""Telnet login. Can be username/password or just password."""
	delay_factor = self.select_delay_factor(delay_factor)
	time.sleep(1 * delay_factor)

	output = ''
	return_msg = ''
	i = 1
	while i <= max_loops:
		try:
			output = self.read_channel()
			return_msg += output

			# Search for username pattern / send username
			if re.search(username_pattern, output):
				self.write_channel(self.username + self.TELNET_RETURN)
				time.sleep(1 * delay_factor)
				output = self.read_channel()
				return_msg += output

			# Search for password pattern / send password
			if re.search(pwd_pattern, output):
				self.write_channel(self.password + self.TELNET_RETURN)
				time.sleep(.5 * delay_factor)
				output = self.read_channel()
				return_msg += output
				if (re.search(pri_prompt_terminator, output, flags=re.M)
						or re.search(alt_prompt_terminator, output, flags=re.M)):
					return return_msg

			# Check if proper data received
			if (re.search(pri_prompt_terminator, output, flags=re.M)
					or re.search(alt_prompt_terminator, output, flags=re.M)):
				return return_msg

			self.write_channel(self.TELNET_RETURN)
			time.sleep(.5 * delay_factor)
			i += 1
		except EOFError:
			msg = "Telnet login failed: {}".format(self.host)
			self.disconnect()
			raise NetMikoAuthenticationException(msg)

	# Last try to see if we already logged in
	self.write_channel(self.TELNET_RETURN)
	time.sleep(.5 * delay_factor)
	output = self.read_channel()
	return_msg += output
	if (re.search(pri_prompt_terminator, output, flags=re.M)
			or re.search(alt_prompt_terminator, output, flags=re.M)):
		return return_msg

	msg = "Telnet login failed: {}".format(self.host)
	self.disconnect()
	raise NetMikoAuthenticationException(msg)	
```
and in cisco base...

```python
    def telnet_login(self, pri_prompt_terminator=r'#\s*$', alt_prompt_terminator=r'>\s*$',
                     username_pattern=r"(?:[Uu]ser:|sername|ogin|User Name)",
                     pwd_pattern=r"assword",
                     delay_factor=1, max_loops=20):
        """Telnet login. Can be username/password or just password."""
        delay_factor = self.select_delay_factor(delay_factor)
        time.sleep(1 * delay_factor)

        output = ''
        return_msg = ''
        i = 1
        while i <= max_loops:
            try:
                output = self.read_channel()
                return_msg += output

                # Search for username pattern / send username
                if re.search(username_pattern, output):
                    self.write_channel(self.username + self.TELNET_RETURN)
                    time.sleep(1 * delay_factor)
                    output = self.read_channel()
                    return_msg += output

                # Search for password pattern / send password
                if re.search(pwd_pattern, output):
                    self.write_channel(self.password + self.TELNET_RETURN)
                    time.sleep(.5 * delay_factor)
                    output = self.read_channel()
                    return_msg += output
                    if (re.search(pri_prompt_terminator, output, flags=re.M)
                            or re.search(alt_prompt_terminator, output, flags=re.M)):
                        return return_msg

                # Support direct telnet through terminal server
                if re.search(r"initial configuration dialog\? \[yes/no\]: ", output):
                    self.write_channel("no" + self.TELNET_RETURN)
                    time.sleep(.5 * delay_factor)
                    count = 0
                    while count < 15:
                        output = self.read_channel()
                        return_msg += output
                        if re.search(r"ress RETURN to get started", output):
                            output = ""
                            break
                        time.sleep(2 * delay_factor)
                        count += 1

                # Check for device with no password configured
                if re.search(r"assword required, but none set", output):
                    msg = "Telnet login failed - Password required, but none set: {}".format(
                        self.host)
                    self.disconnect()
                    raise NetMikoAuthenticationException(msg)

                # Check if proper data received
                if (re.search(pri_prompt_terminator, output, flags=re.M)
                        or re.search(alt_prompt_terminator, output, flags=re.M)):
                    return return_msg

                self.write_channel(self.TELNET_RETURN)
                time.sleep(.5 * delay_factor)
                i += 1
            except EOFError:
                msg = "Telnet login failed: {}".format(self.host)
                self.disconnect()
                raise NetMikoAuthenticationException(msg)

        # Last try to see if we already logged in
        self.write_channel(self.TELNET_RETURN)
        time.sleep(.5 * delay_factor)
        output = self.read_channel()
        return_msg += output
        if (re.search(pri_prompt_terminator, output, flags=re.M)
                or re.search(alt_prompt_terminator, output, flags=re.M)):
            return return_msg

        msg = "Telnet login failed: {}".format(self.host)
        self.disconnect()
        raise NetMikoAuthenticationException(msg)
```

**Comment by ktbyers** [comment]
Fixed in #835
